pysnark/runtime.py

```diff
@@ -483,99 +546,169 @@ def __trunc__(self): return NotImplemented
     __lt__, etc on bitlength-length values
     """
     def check_positive(self, bits=None):
-        if bits is None: bits=bitlength
-            
-        if (is_guard()) and self.value.bit_length()<=bits:
-            ret = PrivVal(1 if self.value>=0 else 0)
-            abs = self.value if self.value>=0 else -self.value
-            bits = [PrivVal((abs&(1<<ix))>>ix) for ix in range(bits)]
+        """
+        Checks if a LinComb is a positive bitlength-bit integer
+        Costs 2 * bitlength + 3 constraints
+        """
+        from pysnark.boolean import PrivValBool
+
+        if bits is None:
+            bits = bitlength
+
+        if is_guard() and self.value.bit_length() <= bits:
+            ret = PrivValBool(1 if self.value >= 0 else 0)
+            abs = self.value if self.value >= 0 else -self.value
+
+            bits = [PrivValBool((abs & (1 << ix)) >> ix) for ix in range(bits)]
         elif ignore_errors():
-            ret = PrivVal(0)
-            bits = [PrivVal(0) for _ in range(bits)]
+            ret = PrivValBool(0)
+            bits = [PrivValBool(0) for _ in range(bits)]
         else:
-            raise ValueError("value " + str(self.value) + " is not a " + str(bits) + "-bit integer")
-            
-        for bit in bits: bit.assert_bool_unsafe()
+            raise ValueError(str(self.value) + " is not a " + str(bits) + "-bit integer")
         
-        # if ret==1, then requires that 2*self=self+sum, so sum=self
-        # if ret==0, this requires that 0=self+sum, so sum=-self
-        add_constraint(2*ret, self, self+LinComb.from_bits(bits))
+        # If ret == 1, then requires that 2 * self = self + sum, so sum = self
+        # If ret == 0, this requires that 0 = self + sum, so sum = -self
+        add_constraint(2 * ret, self, self + LinComb.from_bits(bits))
         
         return ret
             
     def assert_positive(self, bits=None, err=None):
-        if bits is None: bits=bitlength
-            
-        if (not ignore_errors()) and (self.value<0 or self.value.bit_length()>bits):
-            raise ValueError(err if err is not None else "value " + str(self.value) + " is not a " + str(bits) + "-bit positive integer")
+        """
+        Ensures a LinComb is a positive bitlength-bit integer
+        Costs 2 * bitlength + 1 constraints
+        """
 
-        self.to_bits(bits)
+        if bits is None:
+            bits = bitlength
+
+        if not ignore_errors():
+            if self.value < 0 or self.value.bit_length() > bits:
+                raise AssertionError(err if err is not None else str(self.value) + " is not a " + str(bits) + "-bit positive integer")
+
+        self.to_bits()
         
     def check_zero(self):
-        ret = PrivVal(1 if self.value==0 else 0)
-        
-        wit = PrivVal(backend.fieldinverse(self.value+ret.value))  # add ret.value so always nonzero
+        """
+        Checks whether a LinComb is zero
+        Costs 4 constraints
```

Comment:
This costs two constraints, right?

Reply:
This costs 3 constraints:
2 explicit constraints from `add_constraint`
1 constraint from the `LinCombBool` constraint

I fixed the comment in a future commit

Reply:
Ah, but then it could be implemented with two constraints? LinCombBool doesn't need to check again that this is a bit since this is already known?

Reply:
I see. Fixed!